feat: mark `Iff.rfl`, `Iff.symm` proofs for dsimp

thorimur opened the pull request:
This PR modifies the behavior of `dsimp` to allow it to use `Iff.rfl`, `Iff.refl`, and `Iff.symm` proofs of `Iff`. It does so by marking theorems with such proofs as `rfl` theorems within `isRflProof`.

[Needs tests.]

* [X] Put an X in this bracket to confirm you have read the
  [External Contribution Guidelines](https://github.com/leanprover/lean4/blob/master/doc/contributions.md).

* Please put the link to your `RFC` or `bug` issue here.
  PRs missing this link will be marked as `missing RFC`.

Fixes #2678.

* If that issue does not already have approval from a developer,
  please be sure to open this PR in "Draft" mode.

* Please make sure the PR has excellent documentation and tests.
  If we label it `missing documentation` or `missing tests` then it needs fixing!

* You can manage the `awaiting-review`, `awaiting-author`, and `WIP` labels
  yourself, by writing a comment containing one of these labels on its own line.

thorimur commented:
WIP

semorrison commented:
@thorimur, could you run `!bench` on the Mathlib branch? You will need to make a PR out of the branch lean-pr-testing-2679.

semorrison commented:
!bench

thorimur commented:
> @thorimur, could you run `!bench` on the Mathlib branch? You will need to make a PR out of the branch lean-pr-testing-2679.

Will do—as soon as I make sure this is working! :) Turns out I needed to account for `simp`'s preprocessing adding `propext` to these lemmas. (I also plan to add some tests.) Depending on how long it takes to compile and produce a toolchain, I might not have a chance to test and bench until tomorrow.

thorimur commented:
!bench

eric-wieser commented:
> Will do—as soon as I make sure this is working! :) Turns out I needed to account for `simp`'s preprocessing adding `propext` to these lemmas.

What do you mean by this? IIRC, the strategy in lean 3 was just to recreate a copy of the lemma with `=` instead of `Iff`, to avoid needing `propext`.

thorimur commented:
> > Will do—as soon as I make sure this is working! :) Turns out I needed to account for `simp`'s preprocessing adding `propext` to these lemmas.
> 
> What do you mean by this? IIRC, the strategy I implemented in lean 3 was just to recreate a copy of the lemma with `=` instead of `Iff`, to avoid needing `propext`. The motivation is not to avoid the axiom per se, but to ensure everything goes through definitionally, which avoids problems with dependent rewrites.

@eric-wieser The reasons lie in the code `mkSimpTheoremsFromConst` in `Lean.Meta.Tactic.Simp.SimpTheorems`. This calls `preprocess` on the type and value, which in the case of `Iff`, attaches `propext`. In `preprocess` in that same file:
```lean
  ...
  else if let some (lhs, rhs) := type.iff? then
      ...
      let type ← mkEq lhs rhs
      let e    ← mkPropExt e
      return [(e, type)]
```
It is this `e` and `type` which is used to create an `auxLemma`, which is in turn fed to `mkSimpTheoremsCore`—which in turn calls `isRflProof` to set the `dsimp` flag. At that stage, we therefore need to unwrap the `propext` to check if it's `dsimp`-worthy. (I was reluctant to change the flow and wanted to keep my changes minimal, but one could imagine e.g. including the output of `isRflProof` in the `preprocess` result.)

I want to be clear, though, that `isRflProof` *already* unwraps this `auxLemma` anyway, and unwraps constants it encounters recursively (as long as they're applied to something, for some reason). This isn't something I introduced; I just added an `if` branch to handle the `propext` case.

What sorts of problems with dependent rewrites might we worry about here?

semorrison commented:
@thorimur, you might want to rebase this onto `master`; I think we won't be able to get the Mathlib testing to work otherwise, as other things have moved on in the meantime! :-)

nomeata commented:
What is the status of this PR, @thorimur?
